sky/backends/cloud_vm_ray_backend.py

```diff
@@ -363,7 +362,11 @@ def add_ray_task(
         log_path = os.path.expanduser({log_path!r})
 
         if script is not None:
-            sky_env_vars_dict['SKY_NODE_RANK'] = {gang_scheduling_id!r}
+            if {ip_rank_map!r}:
+                ip = ip_list[{gang_scheduling_id!r}]
+                sky_env_vars_dict['SKY_NODE_RANK'] = {ip_rank_map!r}[ip]
```

Comment:
Let's say the cluster has 4 nodes, then `ip_list` has length 4. Let's say this task requires 2 nodes (`sky exec --num-nodes=2 ...`). Then `ip` on L366 may actually map to rank 2 or 3 on L367.  This seems unexpected, as the contract is `SKY_NODE_RANK` should be `{0, 1}`.

Reply:
You are right, thank you for catching this. I've updated the logic as follows: we first create a stable list of all nodes in the cluster `cluster_ips_sorted` (the head node is always first, and the remaining worker nodes are sorted lexicographically). Then, for the list of node IPs that are actually used for the task (which may be subset of the the cluster nodes), we sort this list based on their position in the stable list. This results in the expected behavior.